### PR TITLE
Predicate fn

### PR DESCRIPTION
**Pull request by jvanbruegge**
As discussed in #611 

- [x] I added new tests for the issue I fixed/built
- [x] I ran `make test FOO` for the package FOO I'm modifying
- [x] I used `make commit` instead of `git commit`

### CODE REVIEW

**Comment by staltz** [comment]
Looking good, @jvanbruegge! I'll make some comments inline

File: dom/test/browser/src/events.ts [hunk 1]
```diff
@@ -873,4 +873,70 @@ describe('DOMSource.events()', function () {
     run();
   });
 
+  it('should allow preventing default event behavior with function', function (done) {
```
Comment (staltz) [inline dom/test/browser/src/events.ts]: Would be good to have tests also for the *opposite* case. If the implementation is `if (true) ev.preventDefault()` then these tests would pass.

**Comment by staltz** [comment]
Good. I'll also ask @aronallen's opinion on the object-based API because it's better to get it right in the beginning than to issue a breaking change later on, requiring people to migrate code.

**Comment by aronallen** [comment]
@staltz looks fine for now, the ramda approach may cover us better in the future, but I think this is much simpler and will work in 99% of cases.

File: dom/src/fromEvent.ts [hunk 2]
```diff
@@ -30,3 +33,38 @@ export function fromEvent(
...
+): void {
+  if (preventDefault) {
+    if (typeof preventDefault === 'boolean') {
+      event.preventDefault();
+    } else if (typeof preventDefault === 'function') {
+      if (preventDefault(event)) {
+        event.preventDefault();
+      }
+    } else if (typeof preventDefault === 'object') {
+      const matchObject: (m: {}, o: {}) => boolean = (matcher, obj) =>
```
Comment (staltz) [inline dom/src/fromEvent.ts]: What's the purpose of the matchObject local function if it's only called in one place (just below its declaration)?

Comment (jvanbruegge) [inline dom/src/fromEvent.ts]: it's used twice, most noticably recursive here: https://github.com/cyclejs/cyclejs/pull/615/files/49431069da3dab1759b22c0253cf35949d2c69dc#diff-4974c1e7e3b97190ed437163ebe8d3adR54

Comment (jvanbruegge) [inline dom/src/fromEvent.ts]: Without this recursive matcher, something like `{ ev: { target: { value: 'something' }}}` would not be possible

Comment (aronallen) [inline dom/src/fromEvent.ts]: would this work for arrays as well?
This is a bit of a thought example, but something like:
```javascript
{ target : { classList: ['a', 'b'] } }
```

Comment (jvanbruegge) [inline dom/src/fromEvent.ts]: No, currently it does only an equality check, but i will update this PR later to allow this

Comment (jvanbruegge) [inline dom/src/fromEvent.ts]: @aronallen Done

**Comment by jvanbruegge** [comment]
@staltz is this ready now?

**Comment by staltz** [comment]
Yes, looks all good to merge. 11 days ago when you asked this I was quite busy with conferences. If you don't mind, I'll rebase/squash some commits because `feat` commits will be shown in the changelog and also some of them used `feat(DOM)` instead of `feat(dom)`.

**Comment by jvanbruegge** [comment]
Yeah, go ahead

**Comment by staltz** [comment]
Actually, @jvanbruegge, I think there are still two things to consider: tests for the array part of `match`, I think it was working correctly for arrays even before the commit "feat(DOM): Check arrays on recursive matcher", because `Object.keys(['a', 'b'])` gives `["0", "1"]`. Also, `match` function can be pulled out and become a top-level function. This will actually have better performance because currently every call to `preventDefaultConditional()` will create a **new** `match` function, and functions are objects, so we're spending memory here.

**Comment by jvanbruegge** [comment]
A problem is that I lost my Laptop, so I have no idea when I could change that.

**Comment by staltz** [comment]
Okay, I'll do it then, no stress. And sad to hear you lost your laptop :(
Hope you recover it, or the data, or a new laptop.

**Comment by staltz** [comment]
See PR #678 as a continuation to this.

**Pull request closed without merging**